src/backend/columnar/columnar_customscan.c

```diff
@@ -514,38 +943,68 @@ CreateColumnarScanPath(PlannerInfo *root, RelOptInfo *rel, RangeTblEntry *rte)
+
 	/*
-	 * We don't support pushing join clauses into the quals of a seqscan, but
-	 * it could still have required parameterization due to LATERAL refs in
-	 * its tlist.
+	 * Now, baserestrictinfo contains the clauses referencing only this rel,
+	 * and ppi_clauses (if present) represents the join clauses that reference
+	 * this rel and rels contained in paramRelids (after accounting for
+	 * ECs). Combine the two lists of clauses, extracting the actual clause
+	 * from the rinfo, and filtering out pseudoconstants and SAOPs.
 	 */
-	path->param_info = get_baserel_parampathinfo(root, rel,
-												 rel->lateral_relids);
+	List *allClauses = copyObject(rel->baserestrictinfo);
+	if (path->param_info != NULL)
+	{
+		allClauses = list_concat(allClauses, path->param_info->ppi_clauses);
+	}
 
 	/*
-	 * Add cost estimates for a columnar table scan, row count is the rows estimated by
-	 * postgres' planner.
+	 * This is the set of clauses that can be pushed down for this
+	 * parameterization (with the given paramRelids), and will be used to
+	 * construct the CustomScan plan.
 	 */
-	path->rows = rel->rows;
-	path->startup_cost = 0;
+	List *pushdownClauses = FilterPushdownClauses(root, rel, allClauses);
+
+	if (EnableColumnarQualPushdown)
+	{
+		cpath->custom_private = extract_actual_clauses(pushdownClauses, false);
+	}
+
 	int numberOfColumnsRead = bms_num_members(rte->selectedCols);
-	path->total_cost = path->startup_cost +
-					   ColumnarScanCost(rel, rte->relid, numberOfColumnsRead);
 
-	return (Path *) cpath;
+	CostColumnarScan(cpath, root, rel, rte->relid, numberOfColumnsRead,
+					 list_length(cpath->custom_private));
+
+	add_path(rel, path);
 }
 
 
 /*
- * ColumnarScanCost calculates the cost of scanning the columnar table. The cost is estimated
- * by using all stripe metadata to estimate based on the columns to read how many pages
- * need to be read.
+ * CostColumnarScan calculates the cost of scanning the columnar table. The
+ * cost is estimated by using all stripe metadata to estimate based on the
+ * columns to read how many pages need to be read.
  */
-static Cost
-ColumnarScanCost(RelOptInfo *rel, Oid relationId, int numberOfColumnsRead)
+static void
+CostColumnarScan(CustomPath *cpath, PlannerInfo *root, RelOptInfo *rel,
+				 Oid relationId, int numberOfColumnsRead, int nClauses)
 {
-	return ColumnarTableStripeCount(relationId) *
-		   ColumnarPerStripeScanCost(rel, relationId, numberOfColumnsRead);
+	Path *path = &cpath->path;
+
+	Selectivity clauseSel = clauselist_selectivity(
+		root, cpath->custom_private, rel->relid, JOIN_INNER, NULL);
+
+	/*
+	 * We already filtered out clauses where the overall selectivity would be
+	 * misleading, such as inequalities involving an uncorrelated column. So
+	 * we can apply the selectivity directly to the number of stripes.
+	 */
+	double stripesToRead = clauseSel * ColumnarTableStripeCount(relationId);
+	stripesToRead = Max(stripesToRead, 1.0);
```

Comment:
For index scan, we set `startup_cost` to the cost of reading exactly one stripe.
For `total_cost`, we calculate the cost based on whatever number stripes we will read (i.e. can be 0.25, 1.40 or 56 ...).

For this reason, I got confused why we do that here ? Or, which one is better ?

Reply:
Hm. I'm inclined to say that reading the first stripe shouldn't be part of the startup cost. Startup cost is more for something like HashJoin, or Sort, where it has to do a substantial amount of work before it can return a single tuple.

Reply:
Agree. Then I think we should also do the very same for index scan ? (not add single-stripe read cost to startup-cost, but do `stripesToRead = Max(stripesToRead, 1.0)` when calculating additional total cost)

Reply:
Should we do this in this pr or in another one ?

Reply:
A different PR. This one is taking on too large of a scope, and we'll need to circle back on some cleanup/refactoring.

Reply:
Makes sense, then let's not resolve this comment so that we don't forget about it.
We can also address https://github.com/citusdata/citus/pull/5172#discussion_r692158762 in that other pr.